### ISSUE TITLE
skywalking docker镜像构建k8s部署 | KL博客

### ISSUE THREAD

**Issue by klboke** [klboke/klblog-talk#172]
http://www.kailing.pub/article/index/arcid/221.html 

 KL博客，kl博客，java,php学习心得技术分享，一个爱折腾的程序猿

**Comment by klboke** [comment]
https://www.oschina.net/project/top_cn_2018 skywalking正在参加2018最受欢迎开源项目投票（第四行第一个），非常不错的开源项目，请多多支持

**Comment by klboke** [comment]
docker默认的网络模式是bridge模式，这种模式下，容器里会虚拟出来一个网卡。和主机ip肯定不一样的。可以先了解下docker的网络模型

**Comment by heli-boop** [comment]
按我自己的理解skywalking的这两个配置
agent_gRPC:
  gRPC:
    #OS real network IP(binding required), for agent to uplink data(trace/metrics) to collector. agent--(gRPC)--> collector
    host: localhost
    port: 11800
    # Set these two setting to open ssl
    #sslCertChainFile: $path
    #sslPrivateKeyFile: $path

    # Set your own token to active auth
    #authentication: xxxxxx
agent_jetty:
  jetty:
    # OS real network IP(binding required), for agent to uplink data(trace/metrics) to collector through HTTP. agent--(HTTP)--> collector
    # SkyWalking native Java/.Net/node.js agents don't use this.
    # Open this for other implementor.
    host: localhost
    port: 12800
    contextPath: /

这两个配置的host 是给agent所在应用服务传数据到collecor的，我现在collector服务是docker部署在a机器上，我的agent的服务在b机器上 所以上面的两个localhost被我改成collector容器所在的宿主机的ip了

**Comment by heli-boop** [comment]
http://kailing.pub/article/index/arcid/221.html 你的这篇文章的里面的application.yml里面的realhost变量是可以写成容器所在宿主机的ip吗 还是怎么写 不太理解

**Comment by klboke** [comment]
> @yunqianglee
> 按我自己的理解skywalking的这两个配置
> agent_gRPC:
>   gRPC:
>     #OS real network IP(binding required), for agent to uplink data(trace/metrics) to collector. agent--(gRPC)--> collector
>     host: localhost
>     port: 11800
>     # Set these two setting to open ssl
>     #sslCertChainFile: $path
>     #sslPrivateKeyFile: $path
> 
>     # Set your own token to active auth
>     #authentication: xxxxxx
> agent_jetty:
>   jetty:
>     # OS real network IP(binding required), for agent to uplink data(trace/metrics) to collector through HTTP. agent--(HTTP)--> collector
>     # SkyWalking native Java/.Net/node.js agents don't use this.
>     # Open this for other implementor.
>     host: localhost
>     port: 12800
>     contextPath: /
> 
> 这两个配置的host 是给agent所在应用服务传数据到collecor的，我现在collector服务是docker部署在a机器上，我的agent的服务在b机器上 所以上面的两个localhost被我改成collector容器所在的宿主机的ip了

但是你现在是在容器里面启动的skywalking，所以必须绑定容器里的真实ip，或者0.0.0.0

**Comment by klboke** [comment]
> @yunqianglee
> http://kailing.pub/article/index/arcid/221.html 你的这篇文章的里面的application.yml里面的realhost变量是可以写成容器所在宿主机的ip吗 还是怎么写 不太理解

   - name: real_host
          valueFrom:
            fieldRef:
              fieldPath: status.podIP
在使用k8s部署时，pod里的ip，也就是docker容器里的ip。可以通过以上方式获取并设置为容器的环境变量

**Comment by heli-boop** [comment]
恩 你的说是写成容器自己的ip而非是宿主机的ip，那我不明白的是如果我写成容器的ip ，我另外一台机器的agent服务（也是docker部署）如何访问skywalking的collector服务吗 毕竟他们两个无法直接通信；我认为agent_gRPC的配置是通过某个接口返回给agent服务的，这样agent就知道collector的接口地址上传数据了 这是我目前理解的 不知道哪里出了问题 谢谢回复

**Comment by klboke** [comment]
> @yunqianglee
> 恩 你的说是写成容器自己的ip而非是宿主机的ip，那我不明白的是如果我写成容器的ip ，我另外一台机器的agent服务（也是docker部署）如何访问skywalking的collector服务吗 毕竟他们两个无法直接通信；我认为agent_gRPC的配置是通过某个接口返回给agent服务的，这样agent就知道collector的接口地址上传数据了 这是我目前理解的 不知道哪里出了问题 谢谢回复

如果是k8s容器环境，pod和pod之间是可以直接通讯的，非k8s环境直接docker部署，也可以通过宿主的端口映射到容器内的端口来提供服务的。但是通过服务发现的接口（10800提供）找到的collector服务（11800）确实是容器内的ip+端口，这个时候网络是不通的，也就是我上文结尾谈到的我获取到的是0.0.0.0:11800。skywalking5.x是这样。在skywalking6.x中已经不是这样了，agent直接连接的collector服务，6.x中就没有10800的服务了

**Comment by heli-boop** [comment]
对 所有我在部署collector的时候 加了一个add host=localhost:ip（这个ip就是宿主机的ip） ，这样的话通过服务发现的话 就会返回ip（宿主机的ip）+11800，那么我agent.config  里面写这个ip，然后部署在另外一台机器 应该是可以访问的 但是现在却不行 我也不知道哪里理解的不对 还望解答

**Comment by klboke** [comment]
> @yunqianglee
> 对 所有我在部署collector的时候 加了一个add host=localhost:ip（这个ip就是宿主机的ip） ，这样的话通过服务发现的话 就会返回ip（宿主机的ip）+11800，那么我agent.config  里面写这个ip，然后部署在另外一台机器 应该是可以访问的 但是现在却不行 我也不知道哪里理解的不对 还望解答

这个是行不通的，你的这个想法我在某一瞬间也想到过，但是马上就否了。是因为netty需要绑定ip。必须是当前环境真实ip（容器内ip），或者127.0.0.1，或者是0.0.0.0。

**Comment by heli-boop** [comment]
非常感谢，明白了
